### PR TITLE
Fix stork:// gRPC clients not recovering when the service starts after the client

### PR DESCRIPTION
**Pull request by krickert**
Both Stork-backed gRPC client implementations could get permanently stuck when the target service had no instances at the time of the client's first name resolution — for example when the service is started after the client. Restarting the client was the only way to recover.

For the classic Java gRPC client, `GrpcStorkServiceDiscovery` reported neither a result nor an error to gRPC when Stork returned an empty instance list (its change-detection guard treats empty-to-empty as "nothing changed"). gRPC only re-runs a `NameResolver` after an `onResult`/`onError` callback, so the channel stayed in `CONNECTING` with no scheduled re-resolution.

For the Vert.x gRPC client, `StorkGrpcChannel` built the `DelayedClientCall` with the channel executor rather than the call's own executor. A blocking stub parks the calling thread on a per-call `ThreadlessExecutor`; the close callback after a failed Stork resolution was delivered on the wrong executor, so the call hung instead of failing fast and the client had no chance to retry.

A new `grpc-stork-recovery` integration test reproduces both scenarios (classic and Vert.x clients) using a Stork service discovery that reports no instances until shortly after startup.

### CODE REVIEW

**Comment by krickert** [comment]
To give some additional context of how I discovered this:

I am testing between both types of gRPC clients. My application is a bidi stream that uses consul discovery. 

I created some logic that automatically should reconnect if the server request is down. 

If the server is down when I start my application client, the reconnection never worked. Attempts to reconnect failed.

**Comment by krickert** [comment]
@cescoffier can you look at this one?  This is a significant discovery bug that allows for quarkus to auto-recover (which appears to be designed to do this).. but would love your feedback.

**Comment by krickert** [comment]
Thanks!  If my services start and a dependent service has not yet, then the stork-managed channel never recovers.  This patch will fix that scenario.

**Comment by cescoffier** [comment]
Can you rebase to avoid the merge commit?

File: extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/stork/GrpcStorkServiceDiscovery.java
Comment (cescoffier) [inline extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/stork/GrpcStorkServiceDiscovery.java]: There is a bug here, too.
We need to handle the failure; "resolving" can be stuck.

Comment (cescoffier) [reply extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/stork/GrpcStorkServiceDiscovery.java]: probably something like:

   serviceInstances.subscribe().with(this::informListener, error -> {
       resolving = false;
       log.warnf(error, "Stork service discovery failed for '%s'", serviceName);
       listener.onError(Status.UNAVAILABLE
               .withDescription("Service discovery failed for: " + serviceName)
               .withCause(error));

File: extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/stork/GrpcStorkServiceDiscovery.java [hunk 1]
```diff
@@ -105,6 +105,21 @@ public void refresh() {
...
+                        // No instances were resolved for the service. This must always be
+                        // reported to the listener as an error - even when the previous
+                        // resolution was also empty - because gRPC only re-runs a
+                        // NameResolver (refresh() with backoff) after an onResult or
+                        // onError callback. Staying silent here leaves the channel wedged
+                        // in CONNECTING with no scheduled re-resolution, so it never
+                        // recovers once instances appear (e.g. the service is started
+                        // after the client).
+                        serviceInstanceIds = new HashSet<>();
+                        log.warnf("No service instances resolved for service-name '%s'", serviceName);
```
Comment (cescoffier) [inline extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/stork/GrpcStorkServiceDiscovery.java]: Not sure it's a warning, maybe info, even debug?

gRPC will retry, and run refresh - resolve - inform multiple times. This log might be a bit noisy.

File: integration-tests/grpc-stork-recovery/src/test/java/io/quarkus/grpc/examples/stork/GrpcStorkRecoveryTestBase.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    GreeterGrpc.GreeterBlockingStub hello;
+
+    @Test
+    void channelRecoversWhenServiceAppearsAfterStartup() {
+        HelloRequest request = HelloRequest.newBuilder().setName("late").build();
+
+        // Without a working re-resolution the client stays broken and this
+        // never succeeds within the timeout.
+        await().atMost(Duration.ofSeconds(30)).ignoreExceptions().untilAsserted(() -> {
+            assertThat(call(request).getMessage()).isEqualTo("Hello late");
```
Comment (cescoffier) [inline integration-tests/grpc-stork-recovery/src/test/java/io/quarkus/grpc/examples/stork/GrpcStorkRecoveryTestBase.java:55]: Should we test that the first call fails? We are only checking the eventual success.

File: integration-tests/grpc-stork-recovery/src/main/resources/application.properties [hunk 3]
```diff
@@ -0,0 +1,21 @@
+quarkus.grpc.clients.hello.host=hello-service
+quarkus.grpc.clients.hello.name-resolver=stork
+quarkus.grpc.clients.hello.port=9001
```
Comment (cescoffier) [inline integration-tests/grpc-stork-recovery/src/main/resources/application.properties:3]: Note for myself and @jponge: this will need to be reworked in the Quarkus 4 branch.

File: integration-tests/grpc-stork-recovery/src/main/resources/application.properties [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+quarkus.stork."hello-service".service-discovery.available-after-ms=2000
+quarkus.stork."hello-service".load-balancer.type=round-robin
+
+quarkus.grpc.server.port=9001
+
+# Vert.x gRPC client variant (activated by VertxGRPCTestProfile): the Vert.x
+# client shares the unified HTTP server, so the server and the address served
+# by the 'delayed' discovery both move to the HTTP test port.
+%vertx.quarkus.grpc.clients.hello.use-quarkus-grpc-client=true
+%vertx.quarkus.grpc.clients.hello.port=8081
```
Comment (cescoffier) [inline integration-tests/grpc-stork-recovery/src/main/resources/application.properties]: this could be replaced with: ${quarkus.http.test-port}

**Review by cescoffier — CHANGES_REQUESTED** [review]

File: integration-tests/grpc-stork-recovery/src/main/resources/application.properties (hunk 4 above)
Comment (krickert) [reply integration-tests/grpc-stork-recovery/src/main/resources/application.properties]: done

File: extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/stork/GrpcStorkServiceDiscovery.java (hunk 1 above)
Comment (krickert) [reply extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/stork/GrpcStorkServiceDiscovery.java]: debug now

File: extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/stork/GrpcStorkServiceDiscovery.java
Comment (krickert) [reply extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/stork/GrpcStorkServiceDiscovery.java]: done

**Comment by krickert** [comment]
@aureamunoz and @cescoffier , I assume the failed tests are not related to this change right?

**Comment by aureamunoz** [comment]
> @aureamunoz and @cescoffier , I assume the failed tests are not related to this change right?

No, it doesn't seem to have anything to do with it

File: extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/stork/GrpcStorkServiceDiscovery.java [hunk 5]
```diff
@@ -105,6 +111,21 @@ public void refresh() {
...
+                        // in CONNECTING with no scheduled re-resolution, so it never
+                        // recovers once instances appear (e.g. the service is started
+                        // after the client).
+                        serviceInstanceIds = new HashSet<>();
+                        log.debugf("No service instances resolved for service-name '%s'", serviceName);
+                        listener.onError(Status.UNAVAILABLE
+                                .withDescription("No instances available for service-name: " + serviceName));
+                        return;
+                    }
                     if (serviceInstanceIds.size() != instances.size() || areServicesRemoved(instances)) {
```
Comment (aureamunoz) [inline extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/stork/GrpcStorkServiceDiscovery.java]: I was looking at informListener() and I noticed something that might be related to the same issue this PR is fixing. When instances is non-empty but identical to the previous resolution (this if condition is false), the code inside it is skipped entirely, so neither onResult nor onError is called. From what I've understood, if neither is called, the channel might get stuck and never try to resolve again? it seems like it could be the same kind of "wedge" problem that the empty-list fix is solving, just for a different case. Does it make sense?

File: extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/stork/GrpcStorkServiceDiscovery.java [hunk 6]
```diff
@@ -94,7 +94,13 @@ private void resolve() {
                 }
                 resolving = true;
                 Uni<List<ServiceInstance>> serviceInstances = serviceDiscovery.getServiceInstances();
-                serviceInstances.subscribe().with(this::informListener);
+                serviceInstances.subscribe().with(this::informListener, error -> {
+                    resolving = false;
```
Comment (aureamunoz) [inline extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/stork/GrpcStorkServiceDiscovery.java]: I think I would move this line after the call to the listener for consistency with the other path where the resolving=false is done in the finally, always after the listener call. WDYT?

Comment (aureamunoz) [inline extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/stork/GrpcStorkServiceDiscovery.java]:
```suggestion
                   if (shutdown) {
                        return;
                    }
                    resolving = false;
```
I noticed that the new error handler and the empty-list branch both call listener.onError() without checking shutdown first. Since these are async, shutdown() could be called while Stork is still resolving and the callback would still fire.

File: extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/stork/GrpcStorkServiceDiscovery.java [hunk 7]
```diff
@@ -105,6 +111,21 @@ public void refresh() {
             private void informListener(List<ServiceInstance> instances) {
                 ArrayList<EquivalentAddressGroup> addresses = new ArrayList<>();
                 try {
+                    if (instances.isEmpty()) {
```
Comment (aureamunoz) [inline extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/stork/GrpcStorkServiceDiscovery.java:116]:
```suggestion
                    if (shutdown){
                         return;
                    }
                    if (instances.isEmpty()) {
```
Same here

**Review by aureamunoz — CHANGES_REQUESTED** [review]
Thank you! I added a few comments

File: extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/stork/GrpcStorkServiceDiscovery.java (hunk 6 above)
Comment (cescoffier) [reply extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/stork/GrpcStorkServiceDiscovery.java]: I agree, in case of a shutdown, we should return immediately.

**Review by cescoffier — CHANGES_REQUESTED** [review]
LGTM - just waiting your fixes for @aureamunoz 's comments.

**Comment by cescoffier** [comment]
(I made it "require changes" just to make sure Aurea's comments are addressed before merging).

File: extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/stork/GrpcStorkServiceDiscovery.java (hunk 5 above)
Comment (krickert) [reply extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/stork/GrpcStorkServiceDiscovery.java]: This PR fixes the silent exit for empty lists; the identical-non-empty skip at line 129 is the same class of problem where the refresh completes without calling `onResult` or `onError`.

The scenario I originally reproduced was startup with no instances (empty → empty), one case of that pattern. I'll fix the unchanged-non-empty path in this PR too by always notifying the listener on non-empty resolutions.

File: extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/stork/GrpcStorkServiceDiscovery.java (hunk 6 above)
Comment (krickert) [reply extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/stork/GrpcStorkServiceDiscovery.java]: Implemented the fix.. thanks for the suggestion, included a test.

Comment (krickert) [reply extensions/grpc/runtime/src/main/java/io/quarkus/grpc/runtime/stork/GrpcStorkServiceDiscovery.java]: Done — moved `resolving = false` to after `listener.onError()` for consistency with `informListener()`. Covered by `refreshAfterDiscoveryFailureNotifiesListener`.

**Comment by cescoffier** [comment]
Can you squash your commits?

**Comment by krickert** [comment]
> Can you squash your commits?

Squashed

**Review by aureamunoz — APPROVED** [review]
Super!! thank you very much

**Review by cescoffier — APPROVED** [review]
Thanks!

**Comment by kathbigra** [comment]
@krickert  I think the grpc-stork-recovery module introduced in this PR causes an intermittent build failure during full builds  

**./mvnw install -Dquickly**
error reading .../target/generated-sources/annotations/io/quarkus/grpc/examples/stork/DelayedConfiguration.java
error reading .../target/generated-sources/annotations/io/quarkus/grpc/examples/stork/DelayedServiceDiscoveryProviderLoader.java

The stork-configuration-generator annotation processor writes DelayedConfiguration.java and DelayedServiceDiscoveryProviderLoader.java into target/generated-sources/annotations/ during the same compilation phase that javac tries to read them, causing a read/write conflict. The module passes when built in isolation but fails intermittently in full builds due to build cache interactions.

If you can confirm this, I can provide a fix for the same. I think this can be done by separating the code generation from compilation.

**Comment by krickert** [comment]
Is the error in 

> @krickert I think the grpc-stork-recovery module introduced in this PR causes an intermittent build failure during full builds
> 
> **./mvnw install -Dquickly** error reading .../target/generated-sources/annotations/io/quarkus/grpc/examples/stork/DelayedConfiguration.java error reading .../target/generated-sources/annotations/io/quarkus/grpc/examples/stork/DelayedServiceDiscoveryProviderLoader.java
> 
> The stork-configuration-generator annotation processor writes DelayedConfiguration.java and DelayedServiceDiscoveryProviderLoader.java into target/generated-sources/annotations/ during the same compilation phase that javac tries to read them, causing a read/write conflict. The module passes when built in isolation but fails intermittently in full builds due to build cache interactions.
> 
> If you can confirm this, I can provide a fix for the same. I think this can be done by separating the code generation from compilation.

I'm not sure - I'll try to reproduce it locally and see.  This is a very edge-case scenario, and maybe I made a test that could make it incompatible?  I'll see what I can do to recreate it.

**Comment by krickert** [comment]
@kathbigra I think I've been able to recreate the scenario I'm creating a PR for this now

**Comment by krickert** [comment]
https://github.com/quarkusio/quarkus/pull/54608 @kathbigra  this should fix it .. ran it 25x to make sure it didn't happen

**Comment by anavarr** [comment]
hm.. This PR actually creates another problem: if the backend becomes unreachable (restart pod for instance) the client will enter a tight loop. 
In this tight loop, new subChannels are created on every `handleResolvedAddresses` call (because the LB builds a fresh TreeMap each time), and those subchannels fail immediately with backoff reset to 0, so it spirals to hundreds of thousands of errors per minute and can crash the client.
In sum, this PR completely bypasses the backoff policy that is defined within the load balancer.

Moreover, the initial test on `instances.isEmpty()` calls `listener.onError()` and schedules a refresh. AFAIK, this is enough for solving the initial problem, we can still skip if the service instances didn't change (this will prevent the error log burst, and connection losses are handled by the load balancer down the line.

**Comment by krickert** [comment]
Honestly, thank you for looking into this more. I really hope we can get this totally stable as I'm avoiding unified server 

If you can recreate any of the issues in a test I'll help fix it.

**Comment by krickert** [comment]
@anavarr @cescoffier

I think we should be careful not to fix this by making the resolver hide updates from the load balancer.

The right fix should separate three concerns cleanly: discovery returns endpoints, gRPC name resolution signals results/errors according to the grpc-java contract, and the load balancer owns subchannel lifecycle/backoff without churn.

The empty-list resolver recovery from #54425 is valid: if Stork returns no instances at startup, `GrpcStorkServiceDiscovery` has to call `onError(...)` so grpc-java schedules re-resolution. Otherwise the channel can stay wedged in `CONNECTING` and never recover when the service appears.

The bug exposed here is that `GrpcLoadBalancerProvider` does not safely reconcile resolver updates. It creates new subchannels for each result, does not properly shut down removed ones, and can leave stale subchannel callbacks affecting channel state. Reverting the resolver to skip unchanged non-empty results may reduce the tight loop, but it leaves the load balancer lifecycle issue unresolved.

I can write a follow-up PR to try to fix this properly, but this is a significant reliability issue in the gRPC + Stork surface. I have been hopping between client implementations chasing this behavior, and the only consistently clean escape hatch so far has been avoiding the Quarkus/Stork path and using vanilla Netty gRPC directly.

I do not think we should paper over this by suppressing resolver callbacks instead of fixing subchannel ownership, reuse, shutdown, and backoff behavior. 

I'll make an attempt - but I feel like more eyes should be on this.  I've made two attempts, so far no dice :(
